alphagov/paas-cf
# [#114000923] Implement self-update pipelines for microbosh and cloudfoundry

[#114000923 self-updating pipelines](https://www.pivotaltracker.com/story/show/114000923)

What?
-----

In order to ensure that we promote changes from paas-cf into environments in a consistent way, we need to ensure that changes to concourse pipelines are applied automatically.

To do so, we will add a initial task in each pipeline to update the pipeline itself using the code from the repository.

Notes:

 - The "environment class" specific variables will be set in `Makefile` file, with some sensible defaults defined in code.  In `dev` environments some parameters can be optionally overwritten using environment variables.

 - Self-update will be enabled by default. 

   To disable it export the variable `SELF_UPDATE_PIPELINES=false make dev` when uploading the pipleines. This way one can apply changes without having to commit and push the code.

 - We won't add logic to lock the pipeline as it is out of the scope for this story.

 - Following the Boy Scouts rule, we try simplify the environment class variable definition in the Makefile.


Dependencies
---------------------

 - docker container for https://github.com/alphagov/paas-docker-cloudfoundry-tools/pull/35  must be merged before this one.
 - The previous container must be created in government-paas dockerhub account and the reference changed in this PR. (https://hub.docker.com/r/governmentpaas/self-update-pipelines/)


How to test this?
---------------------

Deploy the environment, but enabling self-update (which is disabled in dev by default), and commit some change to the pipeline. 

I recommend use a temporary branch for this, and pause the job after `init`:

```
git checkout 114000923_self_updating_pipelines_v2
git checkout -b 114000923_self_updating_pipelines_review_$(USER) # Create a temporary branch

BRANCH=$(git rev-parse --abbrev-ref HEAD) make dev
```

Deploy the environment and observe that in the job `init` the self-update pipeline works and do not do any change.

Then add some changes in the pipeline and commit and push the changes. Rerunning the `init` job should show that the pipeline is updated.

You can do the same for:
 - `destroy-microbosh`, `destroy-cloudfoundry` and `failure-testing`
 - using `ci`, `staging` and `prod` targets (in that case export the variable `URSE_URL=https://deployer.hector.dev.paas.alphagov.co.uk`)

Who?
------

Anyone but @saliceti, @keymon or @HenryTK
